KIPs/kip-149.md

```diff
@@ -0,0 +1,231 @@
+
+## Abstract
+
+This standard defines a unified deployment and management process for system contracts. To effectively manage system contracts, it also introduces a Registry contract that manages all system contracts.
+
+## Motivation
+
+Currently, system contracts are deployed and managed without any defined standards. For example, the `AddressBook` contract was deployed by a bytecode injection at the genesis block with a reserved address, while EOA deployed `TreasuryRebalance`, and its address is set in the chain config. As more system contracts will be deployed in the future, it’s essential to have a standard way to deploy and manage system contracts.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+This standard defines the separation of data and logic contracts in system contracts. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized issue and potential private key loss.
+
+### Definitions
+
+- System contract: A contract that affects protocol. The currently deployed system contracts are as follows: **AddressBook**, **Voting**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **StakingTracker**, **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))
+
+- System contract upgrade: The process of updating an existing logic contract while maintaining a proxy contract. The upgraded logic contract must be compatible with the previous interface.
+
+- System contract replacement: The deployment of a new system contract that is then registered to the Registry using the same name as its predecessor. The new contract effectively deprecates the previous system contract.
+
+### Smart Contracts Overview
+
+The proposed smart contract will be implemented in Solidity and compatible with the Ethereum Virtual Machine (EVM).
+
+The smart contract will have the following features:
+
+1. Registry
+
+    - Register a new system contract with an activation block.
+
+    - Return the state of the system contracts.
+
+2. Proxy
+
+    - Delegate a call to logic contract.
+
+    - Upgrade a logic contract.
+
+#### 1. Registry
+
+The registry should have data for the pre-deployed system contracts as soon as it is deployed. It will be done by state injection, which injects data into the Registry directly using the `state.SetState`. A reference implementation is introduce in [Implementation](#implementation).
+
+#### Interface of Registry
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface Registry {
```

Comment:
The specification must stipulate the storage variables because they are crucial to the hardfork consensus.
See https://github.com/klaytn/klaytn/blob/cd09595fe98c768526089d080731e40787b15dcb/contracts/system_contracts/registry/IRegistry.sol.

Laying out storage variables as part of the contract interface is often used technique. For instance, https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/GovernorBravoInterfaces.sol#L63 has a separate GovernorBravoDelegateStorageV1 contract, to make sure that implementations retain the exact storage layout.

Reply:
Since the attached example is for an upgradeable contract (when there's a future possibility to add storage variables), I think the Registry doesn't need that much. Do you expect that there can be a case that we need to add more variables for the Registry contract after initial deployment?

Reply:
But moving the storage variable declaration to the interface for clarification looks better to me.

Reply:
@hyeonLewis thanks. a few more fixes: (1) `interface` cannot contain variables. make it `abstract contract`. (2) variable 'records' and automatic getter function 'records' collides. remove the function. (3) can we rename 'contractNames' to 'names'? because we have 'getAllNames' getter and argument names 'string name'.